### PR TITLE
[MPS] Match CPU bin-range semantics in the histogram kernel

### LINKED ISSUE #195076: [MPS] histogram counts near-edge out-of-range elements that CPU drops, and computes bin index -1
### Describe the bug

On MPS, `torch.histogram` counts elements that lie just outside the bin range. CPU drops them. The window is about 4e-6 wide on both sides. At the lower edge the same path also computes bin index `-1` and stores through it.

```python
import torch

bins = torch.tensor([0.0, 1.0, 2.0])
for edge, deltas in ((0.0, (-1e-7, -1e-6, -3e-6, -5e-6)), (2.0, (1e-6, 3e-6, 5e-6))):
    for d in deltas:
        x = torch.tensor([0.5, edge + d])
        cpu = torch.histogram(x, bins=bins)[0].tolist()
        mps = torch.histogram(x.to("mps"), bins=bins.to("mps"))[0].cpu().tolist()
        print(f"edge={edge} delta={d:>8}  CPU={cpu}  MPS={mps}")
```

Output on 2.15.0a0+gitf5f7ac4, macOS 26.1, Apple M5:

```
edge=0.0 delta=  -1e-07  CPU=[1.0, 0.0]  MPS=[1.0, 1.0]
edge=0.0 delta=  -1e-06  CPU=[1.0, 0.0]  MPS=[1.0, 1.0]
edge=0.0 delta=  -3e-06  CPU=[1.0, 0.0]  MPS=[1.0, 1.0]
edge=0.0 delta=  -5e-06  CPU=[1.0, 0.0]  MPS=[1.0, 0.0]
edge=2.0 delta=   1e-06  CPU=[1.0, 0.0]  MPS=[1.0, 1.0]
edge=2.0 delta=   3e-06  CPU=[1.0, 0.0]  MPS=[1.0, 1.0]
edge=2.0 delta=   5e-06  CPU=[1.0, 0.0]  MPS=[1.0, 0.0]
```

The disagreement is exactly the rows where `|delta| < 4e-6`, at both edges.

### Where it comes from

`aten/src/ATen/native/mps/kernels/HistogramKernel.metal:45` sets `constexpr auto eps = T(4e-6)` and line 57 widens the accept test on both sides. The CPU kernel has no eps (`cpu/HistogramKernel.cpp:153`), which is the whole disagreement.

Bin selection is not widened to match. For an element below the leftmost edge the local `upper_bound` returns `first`, so line 65 gives `pos = -1`. The clamp at line 84 only folds the top edge down, it never raises a negative `pos`, so line 92 stores at a negative `hist_index`. That lands in the previous thread's slice of `thread_histograms`, and `at::sum_out` folds it into the result, which is what makes the `-1e-6` row read `[1.0, 1.0]`. When thread 0 handles the element the same store goes one element before the allocation: `x = torch.tensor([-1e-6, 0.5])` gives `CPU=[1.0, 0.0] MPS=[1.0, 0.0]`, agreeing only because the increment went out of bounds instead of into a neighboring bin.

The upper edge has the count mismatch but not the out-of-bounds store, because of the existing clamp.

### Why I am asking rather than sending a patch

Clamping the low end the way the high end is clamped (`pos = max(pos, 0)`) stops the out-of-bounds store but does not fix the counts: the element lands in bin 0 and MPS returns `[2.0, 0.0]` where CPU returns `[1.0, 0.0]`. The divergence just moves.

The eps comment says it exists because "minuscule implementation differences between the CPU and MPS's linspace" made elements get "accidentally skipped", so it is compensating for edge values computed slightly differently, not deliberately accepting out-of-range input. If that is right, the fix is either to make MPS compute the same edges CPU does, or to snap an element inside the window onto the nearest edge before an unwidened search. Both change results for anyone relying on the wide window, so I would rather have a maintainer pick the intended behavior first.

Note the eps is a fixed absolute constant: `4e-6` is below one ulp of float32 once the leftmost edge is past about 33, and `T(4e-6)` truncates to 0 for integer dtypes, so the window silently disappears in those cases.

### Other affected entry points

- `torch.histogram(input, bins=Tensor)` and `torch.histogramdd(input, bins=[Tensor, ...])` use `BINARY_SEARCH`, which returns `pos = -1` directly.
- `torch.histogram(input, bins=int, range=...)` and the `histogramdd` equivalent use `LINEAR_INTERPOLATION_WITH_LOCAL_SEARCH`, whose `upper_bound` over the clamped window can also return `pos = -1`.
- `torch.histc` uses plain `LINEAR_INTERPOLATION`, which truncates toward zero, so it is safe unless the range is narrower than `bins * 4e-6`.

### Versions

```
PyTorch version: 2.15.0a0+gitf5f7ac4
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 26.1 (arm64)
Clang version: 17.0.0 (clang-1700.4.4.1)
CMake version: version 4.3.1

Python version: 3.13.5 (main, Jun 12 2025, 12:22:43) [Clang 20.1.4 ] (64-bit runtime)
Python platform: macOS-26.1-arm64-arm-64bit-Mach-O
Is CUDA available: False
Is XPU available: False

CPU: Apple M5

Versions of relevant libraries:
[pip3] numpy==2.4.4
[pip3] optree==0.19.0
[pip3] torch==2.15.0a0+gitf5f7ac4
```

cc @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk @Isalia20

### PR DESCRIPTION
**Pull request by kapellirohith**
Fixes #195076

`histogramdd` widens the bin-range accept test by a fixed `eps = 4e-6` on both sides:

```metal
if (!(element >= (leftmost_edge[dim] - eps) &&
      element <= (rightmost_edge[dim] + eps))) {
```

The CPU kernel it was ported from has no such widening. This is a user-visible behaviour change: elements lying just outside the outermost edges were counted on MPS and are now dropped. Checking against numpy as an independent reference rather than against CPU alone:

```
input [-1e-6, 0., 0.5, 1., 2., 2+1e-6], bins [0., 1., 2.]

before   CPU=[2,2]   MPS=[2,3]   numpy=[2,2]
after    CPU=[2,2]   MPS=[2,2]   numpy=[2,2]
```

MPS was the outlier. Anyone relying on the old behaviour was relying on MPS counting elements that both CPU and numpy exclude. This affects all three bin-selection algorithms, `histc` included.

There is a second consequence. For an element below the leftmost edge `upper_bound` returns `first`, so `pos` is `-1`; the existing clamp only folds the top edge down, so the kernel stores through a negative index — either into the previous thread's slice of `thread_histograms`, or, when the first thread owns the element, outside the accumulator entirely. Both effects also occur for non-contiguous input: a transposed view and a strided slice feeding `histogramdd` gave `[1, 1, 2, 0]` on MPS against `[1, 1, 0, 0]` on CPU, so the wrong counts arrive through the strided index indirection as well as through contiguous input.

## Why remove the widening rather than guard `pos`

On the binary-search path the CPU kernel never checks `pos >= 0` because it does not need to. `leftmost_edge[dim]` is assigned directly from `bin_seq[bin_seq_offset]`, so the accept test and the start of the search compare against the same value; an unwidened `element >= leftmost_edge[dim]` therefore guarantees `upper_bound` advances past the first edge and `pos >= 0`. Widening the accept test is what breaks that, and the rest of the kernel was ported assuming it holds. Removing the widening restores the invariant rather than adding a guard downstream.

Scope of that argument: it covers `BINARY_SEARCH`. The two linear-interpolation paths compute `pos` by division instead, and I have not proved the same property for them.

The `eps` was introduced with the original implementation in #96652 to compensate for suspected `linspace` differences between CPU and MPS, and is not a documented decision. That justification no longer holds for the ranges the widening can affect: across the normal-float ranges I tested, bin edges are bitwise identical between CPU and MPS and `linspace` endpoints are exact on MPS. There is one exception — for denormal ranges such as `range=(0, 1e-38)`, CPU produces `[0, 3.33e-39, 6.67e-39, 1e-38]` while MPS produces all zeros — but a `4e-6` window is nine orders of magnitude too coarse to be compensating for that, and it is a separate problem in `linspace` rather than in the accept test.

The widening is also already inert for most of the type matrix: `T(4e-6)` truncates to `0` for the five integer instantiations, and sits below half the edge spacing for `half` and `bfloat` outside a neighbourhood of zero, so those dtypes have always had the CPU behaviour. This change extends it to `float`.

## Testing

Three OpInfo samples place elements one ulp outside the outer edges, covering all three bin-selection algorithms: an explicit edge tensor for `BINARY_SEARCH`, an explicit `range=` for `LINEAR_INTERPOLATION_WITH_LOCAL_SEARCH`, and a `histc` sample with explicit `min`/`max` for `LINEAR_INTERPOLATION`. On the unfixed kernel each of the three diverges from CPU independently. The `histc` sample is guarded on `dtype.is_floating_point`, since `histc` also runs on integer dtypes.

Neither existing yield in `sample_inputs_histogram` could place an element in the widened window, for different reasons. The int-bins yield passes no `range=`, so the outer edges are the input's own min and max via `at::aminmax(input, 0)` (`HistogramKernel.mm:227`) and the extreme elements land exactly on the edges by construction. The explicit-edges yield draws both the input and the edges from `make_tensor`, where the odds of landing within `4e-6` of an edge are negligible. That is why the existing coverage passed.

Two details are deliberate. The out-of-range elements are not placed first, because an invalid bin index computed for the first element writes outside the accumulator and leaves the result looking correct. And the in-range elements do not share a bin with the target of that invalid write, because the accumulator update is not atomic and the two writes otherwise race, losing the evidence.

The samples run on every backend through the device-generic OpInfo tests. On MPS they are picked up by `TestConsistencyMPS.test_output_match` and `test_output_grad_match`, both of which compare forward output against CPU for these ops, and which fail without the kernel change and pass with it.

### CODE REVIEW

**Comment by Skylion007** [comment]
@claude why isn't this exercised by OpInfo test, should it be?

**Comment by kapellirohith** [comment]
histc diverges too, on float32, which I had not expected from the issue text. On the
unfixed kernel, with elements one ulp outside the range:

```
BINARY_SEARCH      histogram(bins=edges)           CPU=[3,1]  MPS=[3,3]
LI + local search  histogram(bins=2, range=(1,3))  CPU=[3,1]  MPS=[3,3]
LINEAR_INTERP      histc(bins=2, min=1, max=3)     CPU=[3,1]  MPS=[4,2]
```

histc differs because truncation toward zero puts the below-range element in bin 0 and
folds the above-range one into the last bin, so both bins are wrong rather than one.
`test_output_match_histc_mps_float32` fails on the unfixed kernel, so the issue's note
that histc is safe unless the range is narrower than `bins * 4e-6` does not hold — the
range here is 2.0. I have pushed samples for all three paths, each verified to diverge
independently.

I kept the `pos >= 0` argument scoped to the binary search. On the linear paths
`(element - lo) * (n_edges - 1)` overflows to inf for a range like (-1e38, 1e38) even
when the element is in range, so truncation does not save it there. That overflow also
makes CPU and MPS disagree with numpy and with each other, which looks pre-existing and
not MPS-specific. Worth a separate issue, or is it already known?
